### PR TITLE
Make builds faster (sub 7 minutes)

### PR DESCRIPTION
**Pull request by SerialVelocity**
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1379)

### CODE REVIEW

File: atlasdb-client-protobufs/build.gradle [hunk 1]
```diff
@@ -6,7 +6,7 @@ buildscript {
     mavenCentral()
   }
   dependencies {
-    classpath 'com.google.protobuf:protobuf-gradle-plugin:0.7.6'
```
Comment (hsaraogi) [inline atlasdb-client-protobufs/build.gradle]: Did you regenerate the protobufs. All the `*Persisitence.java` files should be regenerated and changes, if any be merged.

Comment (hsaraogi) [inline atlasdb-client-protobufs/build.gradle]: Did you regenerate the .proto files. All the `*.Persistence` files should be regenerated and the changes if any committed along.

File: circle.yml [hunk 2]
```diff
@@ -2,13 +2,13 @@ machine:
   pre:
     - ./scripts/circle-ci/install-docker.sh:
         pwd: $CIRCLE_PROJECT_REPONAME
-    - sudo pip install docker-compose
+    - sudo wget --retry-connrefused --waitretry=1 --read-timeout 20 --timeout 15 -t 10 -q -O /usr/local/bin/docker-compose https://github.com/docker/compose/releases/download/1.9.0/docker-compose-`uname -s`-`uname -m`
```
Comment (hsaraogi) [inline circle.yml]: We are locking to docker-compose 1.9.0. Can we use the latest release always http://stackoverflow.com/questions/24987542/is-there-a-link-to-github-for-downloading-a-file-in-the-latest-release-of-a-repo?

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java [hunk 3]
```diff
@@ -52,6 +53,8 @@
 import io.dropwizard.testing.ResourceHelpers;
 import io.dropwizard.testing.junit.DropwizardAppRule;
 
+@Ignore("Observed ConcurrentModificationException-related flakes (e.g. build #611 on internal CircleCI)."
+        + "Fixed in atomix/copycat#231, but not part of Copycat 1.1.4 which we use.")
 public class TimeLockServerIntegrationTest {
```
Comment (hsaraogi) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java]: Awesome, this makes it very clear why these are ignored.

File: atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra-three-node.yml [hunk 4]
```diff
@@ -5,6 +5,8 @@ services:
     image: palantirtechnologies/docker-cassandra-atlasdb:$CASSANDRA_VERSION
     environment:
       - CASSANDRA_SEEDS=cassandra1,cassandra2,cassandra3
+      - MAX_HEAP_SIZE=$CASSANDRA_MAX_HEAP_SIZE
+      - HEAP_NEWSIZE=$CASSANDRA_HEAP_NEWSIZE
```
Comment (hsaraogi) [inline atlasdb-container-test-utils/src/main/resources/docker-compose-cassandra-three-node.yml]: Just out of curiosity, why do we override these values. If they are recommended values, should we also modify the docker images to be built with this configuration. Also, do you have a (intutive)measure of whether this save us time?

File: circle.yml [hunk 5]
```diff
@@ -17,18 +17,14 @@ checkout:
   post:
     - ./scripts/circle-ci/pull-all-images.sh:
         background: true
-    - ./scripts/circle-ci/stop-services.sh
-
-general:
-  artifacts:
-    - "build/reports/profile"
+    - ./scripts/circle-ci/stop-services.sh:
+        background: true
```
Comment (hsaraogi) [inline circle.yml]: Is it fine if some services are stopped after tests start running? Background doesnt guarantee that this will be complete before. There are two background scripts now, are both run in parallel, pull-images is likely to take a long time.

File: scripts/circle-ci/run-circle-tests.sh [hunk 6]
```diff
@@ -36,12 +39,25 @@ if ./scripts/circle-ci/check-only-docs-changes.sh; then
     exit 0
 fi
 
+if [[ $INTERNAL_BUILD == true ]]; then
+    BASE_GRADLE_ARGS+=" --parallel"
```
Comment (hsaraogi) [inline scripts/circle-ci/run-circle-tests.sh]: Add " --parallel" to BASE_GRADLE_ARGS.

File: circle.yml (hunk 5 above)
Comment (SerialVelocity) [inline circle.yml]: `stop-services.sh` only takes 10s. They should stop in time. Stopping the services is only to reduce memory pressure in external CircleCI so it doesn't really matter if the tests have started.

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java (hunk 3 above)
Comment (SerialVelocity) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java]: I copied it from @jeremyk-91's `@Ignore` xD

File: scripts/circle-ci/run-circle-tests.sh (hunk 6 above)
Comment (SerialVelocity) [inline scripts/circle-ci/run-circle-tests.sh]: We don't want it for external builds since it causes OOMs

File: atlasdb-client-protobufs/build.gradle (hunk 1 above)
Comment (jboreiko) [inline atlasdb-client-protobufs/build.gradle]: Woah, if any of the persisted files changed this cannot merge unless we can confirm that their on disk formats have not changed.  

That said, changing the gradle plugin shouldn't result in any changes to those files.

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java [hunk 7]
```diff
@@ -52,6 +53,8 @@
 import io.dropwizard.testing.ResourceHelpers;
 import io.dropwizard.testing.junit.DropwizardAppRule;
 
+@Ignore("Observed ConcurrentModificationException-related flakes (e.g. build #611 on internal CircleCI)."
```
Comment (jboreiko) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java]: Can we file a ticket to ensure we track unignoring these or removing them.

File: circle.yml [hunk 8]
```diff
@@ -2,13 +2,13 @@ machine:
...
         pwd: $CIRCLE_PROJECT_REPONAME
-    - sudo pip install docker-compose
+    - sudo wget --retry-connrefused --waitretry=1 --read-timeout 20 --timeout 15 -t 10 -q -O /usr/local/bin/docker-compose https://github.com/docker/compose/releases/download/1.9.0/docker-compose-`uname -s`-`uname -m`
+    - sudo chmod +x /usr/local/bin/docker-compose
   java:
     version:
       oraclejdk8
   environment:
     _JAVA_OPTIONS: "-Xmx512M"
-    GRADLE_OPTS: '-Xss1024K -XX:+CMSClassUnloadingEnabled -XX:InitialCodeCacheSize=32M -XX:CodeCacheExpansionSize=1M -XX:CodeCacheMinimumFreeSpace=1M -XX:ReservedCodeCacheSize=150M -XX:MinMetaspaceExpansion=1M -XX:MaxMetaspaceExpansion=8M -XX:MaxMetaspaceSize=128M -XX:MaxDirectMemorySize=96M -XX:CompressedClassSpaceSize=32M'
```
Comment (jboreiko) [inline circle.yml]: Why did we remove these?

File: circle.yml [hunk 9]
```diff
@@ -60,6 +54,8 @@ test:
         parallel: true
     - find . -type d -regex ".*/build/reports/findbugs" | sed 's#\./\(\(.*\)/build/reports/findbugs\)#rsync -uav \1/ $CIRCLE_ARTIFACTS/findbugs/\2#' | bash:
         parallel: true
+    - rsync -uav "build/reports/profile" "$CIRCLE_ARTIFACTS/":
```
Comment (jboreiko) [inline circle.yml]: rsync over mv?

File: gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -1,6 +1,6 @@
-#Tue Aug 16 17:01:34 BST 2016
+#Tue Dec 13 19:32:41 GMT 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.2.1-bin.zip
```
Comment (jboreiko) [inline gradle/wrapper/gradle-wrapper.properties]: Dope :)

File: circle.yml (hunk 8 above)
Comment (jboreiko) [inline circle.yml]: ah, see now that we only do it for external.

File: scripts/circle-ci/run-circle-tests.sh [hunk 11]
```diff
@@ -2,25 +2,28 @@
```
Comment (jboreiko) [inline scripts/circle-ci/run-circle-tests.sh]: Might be worth including a comment about the differences between internal and external that have resulted in the below separation.  If I were knew, this would be pretty hard to parse.

File: circle.yml (hunk 2 above)
Comment (SerialVelocity) [inline circle.yml]: Talked offline about this. We're going to leave it as 1.9.0. Keeping it at 1.9.0 also gives more reproducible builds.

File: circle.yml (hunk 9 above)
Comment (SerialVelocity) [inline circle.yml]: Gives a status report :)

File: atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java (hunk 7 above)
Comment (SerialVelocity) [inline atlasdb-timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/TimeLockServerIntegrationTest.java]: We're already moving away from atomix (due to correctness issues with it) so these will be deleted soon

File: atlasdb-client-protobufs/build.gradle (hunk 1 above)
Comment (SerialVelocity) [inline atlasdb-client-protobufs/build.gradle]: Checked this. It doesn't regenerate anything. (Our auto-regeneration code has never worked too. I've got a fix but it also generates UpgradePersistence.java which we may want to delete from the repo)

File: scripts/circle-ci/run-circle-tests.sh (hunk 11 above)
Comment (SerialVelocity) [inline scripts/circle-ci/run-circle-tests.sh]: Done

File: build.gradle [hunk 12]
```diff
@@ -27,7 +27,7 @@ plugins {
     id 'com.github.johnrengelman.shadow' version '1.2.3'
     id 'com.palantir.configuration-resolver' version '0.1.0'
     id 'com.palantir.git-version' version '0.5.2'
-    id 'org.inferred.processors' version '1.1'
+    id 'org.inferred.processors' version '1.2.4-rc2'
```
Comment (schlosna) [inline build.gradle]: Why are we using RC build and what repo should that be coming from? Using `1.2.4-rc2` results in the following error for me, while dropping down to `1.2.3` works fine.

```
$ ./gradlew clean idea
:buildSrc:compileJava UP-TO-DATE
:buildSrc:compileGroovy UP-TO-DATE
:buildSrc:processResources UP-TO-DATE
:buildSrc:classes UP-TO-DATE
:buildSrc:jar UP-TO-DATE
:buildSrc:assemble UP-TO-DATE
:buildSrc:compileTestJava UP-TO-DATE
:buildSrc:compileTestGroovy UP-TO-DATE
:buildSrc:processTestResources UP-TO-DATE
:buildSrc:testClasses UP-TO-DATE
:buildSrc:test UP-TO-DATE
:buildSrc:check UP-TO-DATE
:buildSrc:build UP-TO-DATE

FAILURE: Build failed with an exception.

* Where:
Build file '/Volumes/git/atlasdb/build.gradle' line: 30

* What went wrong:
An exception occurred applying plugin request [id: 'org.inferred.processors', version: '1.2.4-rc2']
> Failed to apply plugin [id 'org.inferred.processors']
   > Could not get unknown property 'idea' for root project 'atlasdb' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 1.009 secs
```

Comment (schlosna) [inline build.gradle]: We should probably cut a real `1.2.4` with the changes https://github.com/palantir/gradle-processors/compare/v1.2.3...master

@chrisalice for SA

Comment (chrisalice) [inline build.gradle]: Then again, I can't repro the bug and you've merged this PR. Was it a transient failure?
